app/src/main/java/ciserver/PipelineCompiler.java

```diff
@@ -9,43 +9,43 @@
 import org.apache.commons.io.IOUtils;
 
 public class PipelineCompiler implements StageTask {
-	public final ArrayList<String> commands = new ArrayList<>();
-
-	PipelineCompiler(String... commands) {
-		this.commands.addAll(Arrays.asList(commands));
-	}
-
-	/**
-	 * Run the specified `commands` in the folder of the pulled repository.
-	 *
-	 * Note: the PipelinePull component or a similar action must be run before this.
-	 *
-	 * @return the status of the compilation action
-	 */
-	public PipelineStatus execute(String pipelineDir, PushEvent event) {
-		var path = String.format("%s/repositories/%s", pipelineDir, event.headCommit.id);
-		var directory = new File(path);
-		directory.mkdirs(); // Make the directories recursively
-
-		try {
-			var process = new ProcessBuilder(commands)
-					.directory(directory)
-					.redirectErrorStream(true)
-					.start();
-
-			// Here we have access to the output of the build, which we can send to GitHub.
-			String output = IOUtils.toString(process.getInputStream(), StandardCharsets.UTF_8);
-			System.err.println(output);
-
-			process.waitFor();
-		} catch (IOException e) {
-			e.printStackTrace();
-			return PipelineStatus.Fail;
-		} catch (InterruptedException e) {
-			e.printStackTrace();
-			return PipelineStatus.Fail;
-		}
-
-		return PipelineStatus.Ok;
-	}
+    final ArrayList<String> commands = new ArrayList<>();
```

Comment:
Also final doesn't really do anything here since it just prevents reassignment, but it doesn't hurt so ¯\\\_(ツ)_/¯

```suggestion
    private final ArrayList<String> commands = new ArrayList<>();
```

Reply:
Variables should automatically be immutable, don't try to change my mind about this :^)